<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#📌-Project-Description" data-toc-modified-id="📌-Project-Description-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>📌 Project Description</a></span></li><li><span><a href="#🎯-Objectives" data-toc-modified-id="🎯-Objectives-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>🎯 Objectives</a></span></li><li><span><a href="#⚙️-Initialization" data-toc-modified-id="⚙️-Initialization-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>⚙️ Initialization</a></span></li><li><span><a href="#🔗-Connecting-to-the-Database" data-toc-modified-id="🔗-Connecting-to-the-Database-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>🔗 Connecting to the Database</a></span></li><li><span><a href="#📊-Exploring-the-Tables" data-toc-modified-id="📊-Exploring-the-Tables-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>📊 Exploring the Tables</a></span></li><li><span><a href="#🧪-Exercises" data-toc-modified-id="🧪-Exercises-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>🧪 Exercises</a></span><ul class="toc-item"><li><span><a href="#📘Number-of-books-published-after-January-1,-2000" data-toc-modified-id="📘Number-of-books-published-after-January-1,-2000-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>📘Number of books published after January 1, 2000</a></span></li><li><span><a href="#📗-Number-of-User-Reviews-and-Average-Rating-per-Book" data-toc-modified-id="📗-Number-of-User-Reviews-and-Average-Rating-per-Book-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>📗 Number of User Reviews and Average Rating per Book</a></span></li><li><span><a href="#📘-Publisher-with-the-Most-Books-Over-50-Pages" data-toc-modified-id="📘-Publisher-with-the-Most-Books-Over-50-Pages-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>📘 Publisher with the Most Books Over 50 Pages</a></span></li><li><span><a href="#🖋️-Author-with-the-Highest-Average-Book-Rating-(Min.-50-Ratings-per-Book)" data-toc-modified-id="🖋️-Author-with-the-Highest-Average-Book-Rating-(Min.-50-Ratings-per-Book)-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>🖋️ Author with the Highest Average Book Rating (Min. 50 Ratings per Book)</a></span></li><li><span><a href="#💬-Average-Number-of-Text-Reviews-Among-Users-Who-Rated-More-Than-50-Books" data-toc-modified-id="💬-Average-Number-of-Text-Reviews-Among-Users-Who-Rated-More-Than-50-Books-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>💬 Average Number of Text Reviews Among Users Who Rated More Than 50 Books</a></span></li></ul></li></ul></div>

## 📌 Project Description

The COVID-19 pandemic took the world by surprise, dramatically changing people’s daily routines. City dwellers stopped spending their free time at cafés and shopping centers and instead turned to activities at home — such as reading. This surge in reading habits sparked the interest of startups that quickly launched new platforms tailored for book lovers.

In this project, we are given access to a database from one of these emerging services. The dataset includes information about books, publishers, authors, user ratings, and written reviews. Our goal is to analyze this data to generate insights that will help define a value proposition for a new product in the book-tech market.


## 🎯 Objectives

The following analytical questions will be addressed using SQL:

1. Identify the number of books published after January 1, 2000.
2. Determine the number of user reviews and the average rating for each book.
3. Find the publisher that has released the highest number of books with more than 50 pages.
4. Identify the author with the highest average rating, considering only books with at least 50 ratings.
5. Calculate the average number of text reviews written by users who have rated more than 50 books.

## ⚙️ Initialization

In this section, we load the necessary libraries to connect to the PostgreSQL database and handle the data within the notebook.

In [4]:
# Uncomment the following lines if you haven't installed these packages yet:
# !python -m pip install --upgrade 'sqlalchemy<2.0'
# !pip install psycopg2

In [5]:
import pandas as pd
from sqlalchemy import create_engine
import psycopg2

## 🔗 Connecting to the Database

In this step, we will connect to the PostgreSQL database in order to access the dataset. The goal is to explore the database structure, identify relevant variables, and assess the quality of the data.

In [7]:
# Replace the following with your actual credentials
db_config = {'user': 'practicum_student', # nombre de usuario
'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # puerto de conexión
'db': 'data-analyst-final-project-db'} # nombre de la base de datos

# Create connection string
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],

db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

# Test connection by displaying available tables (optional)
engine = create_engine(connection_string, connect_args={'sslmode':'require'})# Deprecated in SQLAlchemy 2.0+, use inspect(engine).get_table_names() instead if needed


## 📊 Exploring the Tables

Now that the database connection is established, let's take a first look at the data.

We will display the first few rows of each table to understand the structure, column names, and data formats. This will help us plan the queries more effectively and identify any data quality issues early on.


In [8]:
#View first rows from books
query= "SELECT * FROM books LIMIT 5"
pd.read_sql(query, engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [9]:
#View first rows from authors
query= "SELECT * FROM authors LIMIT 5"
pd.read_sql(query, engine)


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [10]:
#View first rows from publishers
query= "SELECT * FROM publishers LIMIT 5"
pd.read_sql(query, engine)


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
#View first rows from ratings
query= "SELECT * FROM ratings LIMIT 5"
pd.read_sql(query, engine)


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
#View first rows from reviews
query= "SELECT * FROM reviews LIMIT 5"
pd.read_sql(query, engine)


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## 🧪 Exercises

### 📘Number of books published after January 1, 2000

We want to identify how many books in the dataset were published after January 1, 2000. This will give us an idea of how much recent content the platform offers, which is important for understanding current trends and user preferences.

In [13]:
query = """
        SELECT COUNT(*) AS num_libros
        FROM books
        WHERE publication_date > '2000-01-01';
        """
pd.read_sql(query, engine)


num_libros
0         819

A total of **819 books** in the dataset were published after January 1, 2000. This suggests that the platform includes a significant number of modern publications, which is important for attracting contemporary readers and staying relevant in the current market.

### 📗 Number of User Reviews and Average Rating per Book

In this step, we want to determine two key metrics for each book:

- How many reviews (ratings) it has received.
- What is its average rating.

This information is useful for identifying books that are popular and well-rated by users — key indicators for recommendation systems and market positioning.

In [14]:
query = """
SELECT
    b.book_id,
    COALESCE(COUNT(r.rating_id), 0) AS num_reseñas,
    COALESCE(AVG(r.rating), 0) AS calificacion_promedio
FROM
    books b
LEFT JOIN
    ratings r ON b.book_id = r.book_id
GROUP BY
    b.book_id;
        """
pd.read_sql(query, engine)


book_id  num_reseñas  calificacion_promedio
0        652            2               4.500000
1        273            2               4.500000
2         51           12               4.250000
3        951            2               4.000000
4        839            7               4.285714
..       ...          ...                    ...
995       64           13               4.230769
996       55            2               5.000000
997      148            7               3.428571
998      790            2               3.500000
999      828            2               3.000000

[1000 rows x 3 columns]

### 📘 Publisher with the Most Books Over 50 Pages

We want to identify which publisher has released the highest number of books that are longer than 50 pages.

This filter helps us exclude short publications like pamphlets or manuals, and instead focus on substantial works. Understanding which publisher contributes the most full-length books may be useful for partnership opportunities or catalog assessment.

In [15]:
query = """
SELECT
    p.publisher_id,
    p.publisher,
    COUNT(CASE WHEN b.num_pages > 50 THEN b.book_id END) AS num_libros
FROM
    books b
JOIN
    publishers p ON b.publisher_id = p.publisher_id
GROUP BY
    p.publisher_id, p.publisher
ORDER BY
    num_libros DESC
LIMIT 1;
"""
pd.read_sql(query, engine)


publisher_id      publisher  num_libros
0           212  Penguin Books          42

The publisher with the highest number of books over 50 pages is **Penguin Books**, with a total of **42 titles**.

This suggests that Penguin Books is a leading contributor of full-length books in the dataset, which may indicate a strong and diverse catalog suitable for readers seeking substantial content.	

### 🖋️ Author with the Highest Average Book Rating (Min. 50 Ratings per Book)

We want to find out which author has the highest average book rating, but only considering books that have received at least 50 user ratings.

This filter helps us avoid bias from books with too few ratings, which could distort the results. This metric can highlight authors who consistently produce high-quality and well-received work.

In [16]:
query = """
SELECT
    a.author_id,
    a.author,
    AVG(r.rating) AS calificacion_promedio
FROM
    authors a
JOIN
    books b ON a.author_id = b.author_id
JOIN
    ratings r ON b.book_id = r.book_id
GROUP BY
    a.author_id, a.author
HAVING
    COUNT(r.rating_id) >= 50
ORDER BY
    calificacion_promedio DESC
LIMIT 1;
"""
pd.read_sql(query, engine)



author_id          author  calificacion_promedio
0        130  Diana Gabaldon                    4.3

The author with the highest average book rating, considering only books with at least 50 ratings, is **Diana Gabaldon**, with an impressive average score of **4.3**.

This indicates that Diana Gabaldon’s works are not only widely read but also consistently well-received by the audience, making her a strong candidate for highlighted recommendations or featured content.	

### 💬 Average Number of Text Reviews Among Users Who Rated More Than 50 Books

This analysis focuses on the most active users — those who have rated more than 50 books — and calculates the average number of written reviews they submitted.

By comparing the number of ratings with the number of written reviews, we gain insight into how engaged these users are beyond just giving stars. This can help assess the quality of user feedback and potential for social features like community reviews or top reviewer badges.

In [17]:
query = """
SELECT
    AVG(n_reviews_text) AS promedio_reseñas_texto
FROM (
    SELECT
        username,
        COUNT(review_id) AS n_reviews_text
    FROM
        reviews
    WHERE
        username IN (
            SELECT
                username
            FROM
                ratings
            GROUP BY
                username
            HAVING
                COUNT(rating_id) > 50
        )
    GROUP BY
        username
) AS subquery;
"""
pd.read_sql(query, engine)

promedio_reseñas_texto
0               24.333333

On average, users who have rated more than 50 books wrote **24.33 text reviews**.

This indicates a fairly high level of engagement among the most active users, not only are they rating many books, but they are also providing written feedback on nearly half of them. This type of user behavior is valuable for platforms aiming to foster community-driven content and personalized recommendations.
